### PR TITLE
Refactoring CMS/PKCS#7 and better use of the memory

### PR DESCRIPTION
**Pull request by wargio**
@Maijin it fixes the `signature.exe` cms data `iC` output (and also any 4096 bits signatures)

### CODE REVIEW

File: libr/bin/format/pe/pe.c [hunk 1]
```diff
@@ -1898,9 +1900,19 @@ static void bin_pe_get_certificate (struct PE_ (r_bin_pe_obj_t) * bin) {
...
 		return;
 	}
-	con = r_pkcs7_parse_container (data, size);
-	bin->pkcs7 = r_pkcs7_container_dump (con);
-	r_pkcs7_free_container (con);
+	con = r_pkcs7_parse_cms (data, size);
+	bin->pkcs7 = r_pkcs7_cms_dump (con);
+	if (bin->pkcs7) {
+		ut64 length = strlen(bin->pkcs7);
+		char *c = (char*) malloc(length);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: missing space before (

File: libr/bin/format/pe/pe.c [hunk 2]
```diff
@@ -1898,9 +1900,19 @@ static void bin_pe_get_certificate (struct PE_ (r_bin_pe_obj_t) * bin) {
...
+		if (c) {
+			memcpy(c, bin->pkcs7, length);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: missing space

File: libr/bin/format/pe/pe.c [hunk 3]
```diff
@@ -1898,9 +1900,19 @@ static void bin_pe_get_certificate (struct PE_ (r_bin_pe_obj_t) * bin) {
...
+			c[length-1] = '\0';
+			free((char*)bin->pkcs7);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: missing space

File: libr/util/r_x509.c [hunk 4]
```diff
@@ -638,17 +641,19 @@ char* r_x509_certificate_dump (RX509Certificate* certificate, char* buffer, ut32
...
+//	r = snprintf (buffer + p, length - p, "%sAlgorithm:\n%s%s\n%sSignature: %u bytes\n%s\n",
+//				pad, pad2, algo ? algo->string : "",
+//				pad, certificate->signature->length, signature ? signature->string : "");
+	r = snprintf (buffer + p, length - p, "%sAlgorithm:\n%s%s\n%sSignature: %u bytes\n",
+				pad, pad2, algo ? algo->string : "", pad, certificate->signature->length);
 	if (r < 0) {
 		free (pad2);
 		return NULL;
 	}
 	p += (unsigned) r;
```
Comment (radare) [inline libr/util/r_x509.c]: use unsigned int or ut32, but not `unsigned`

File: libr/include/r_util/r_pkcs7.h [hunk 5]
```diff
@@ -63,11 +63,11 @@ typedef struct r_pkcs7_signeddata_t {
 typedef struct r_pkcs7_container_t {
 	RASN1String *contentType;
 	RPKCS7SignedData signedData;
-} RPKCS7Container;
+} RCMS;
 
-R_API RPKCS7Container *r_pkcs7_parse_container (const ut8 *buffer, ut32 length);
-R_API void r_pkcs7_free_container (RPKCS7Container* container);
-R_API char* r_pkcs7_container_dump (RPKCS7Container* container);
+R_API RCMS *r_pkcs7_parse_cms (const ut8 *buffer, ut32 length);
```
Comment (radare) [inline libr/include/r_util/r_pkcs7.h]: function signatures doesnt have a space before the (, use r2-indent

File: libr/bin/format/pe/pe.c [hunk 6]
```diff
@@ -1898,9 +1900,19 @@ static void bin_pe_get_certificate (struct PE_ (r_bin_pe_obj_t) * bin) {
...
-	bin->pkcs7 = r_pkcs7_container_dump (con);
-	r_pkcs7_free_container (con);
+	con = r_pkcs7_parse_cms (data, size);
+	bin->pkcs7 = r_pkcs7_cms_dump (con);
+	if (bin->pkcs7) {
+		ut64 length = strlen (bin->pkcs7);
+		char *c = (char*) malloc (length);
+		if (c) {
+			memcpy (c, bin->pkcs7, length);
+			c[length-1] = '\0';
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: spaces around -

File: libr/bin/format/pe/pe.c [hunk 7]
```diff
@@ -1903,12 +1903,12 @@ static void bin_pe_get_certificate (struct PE_ (r_bin_pe_obj_t) * bin) {
 	con = r_pkcs7_parse_cms (data, size);
 	bin->pkcs7 = r_pkcs7_cms_dump (con);
 	if (bin->pkcs7) {
-		ut64 length = strlen(bin->pkcs7);
-		char *c = (char*) malloc(length);
+		ut64 length = strlen (bin->pkcs7);
```
Comment (wargio) [inline libr/bin/format/pe/pe.c]: actually it handles size_t which is a int64_t on 64 bit machine, and int32_t on 32 bit machine

**Pull request merged**